File: lytecache-go/zeroconfig_test.go
package lytecache_test

import (
	"os"
	"path/filepath"
	"testing"

	lytecache "github.com/YOURUSERNAME/lytecache-go"
)

// patchCacheDir redirects os.UserCacheDir to a temp directory for the
// duration of the test, across the platforms Go's UserCacheDir supports:
// HOME (used directly on macOS, and as the ".cache" fallback on Linux),
// XDG_CACHE_HOME (Linux, cleared so the HOME fallback applies), and
// LocalAppData (Windows).
func patchCacheDir(t *testing.T, dir string) {
	t.Helper()
	t.Setenv("HOME", dir)
	t.Setenv("XDG_CACHE_HOME", "")
	t.Setenv("LocalAppData", dir)
}

func TestZeroConfigCreatesFileAndParentDirs(t *testing.T) {
	patchCacheDir(t, t.TempDir())

	path, err := lytecache.DefaultPath()
	if err != nil {
		t.Fatal(err)
	}
	if _, err := os.Stat(path); !os.IsNotExist(err) {
		t.Fatalf("expected no file yet at %s", path)
	}
	if _, err := os.Stat(filepath.Dir(path)); !os.IsNotExist(err) {
		t.Fatalf("expected the parent directory not to exist yet either")
	}

	c, err := lytecache.New()
	if err != nil {
		t.Fatal(err)
	}
	defer func() { _ = c.Close() }()

	if _, err := os.Stat(path); err != nil {
		t.Fatalf("expected New() to create the file: %v", err)
	}
	if c.Path() != path {
		t.Fatalf("expected Path() %q, got %q", path, c.Path())
	}
}

func TestLytecachePathEnvOverride(t *testing.T) {
	dir := t.TempDir()
	override := filepath.Join(dir, "custom", "nested", "path.db")
	t.Setenv("LYTECACHE_PATH", override)

	path, err := lytecache.DefaultPath()
	if err != nil {
		t.Fatal(err)
	}
	if path != override {
		t.Fatalf("expected %q, got %q", override, path)
	}

	c, err := lytecache.New()
	if err != nil {
		t.Fatal(err)
	}
	defer func() { _ = c.Close() }()

	if c.Path() != override {
		t.Fatalf("expected %q, got %q", override, c.Path())
	}
	if _, err := os.Stat(override); err != nil {
		t.Fatalf("expected file to be created at the override path: %v", err)
	}
}

func TestDifferentWorkingDirectoriesResolveDifferentFiles(t *testing.T) {
	patchCacheDir(t, t.TempDir())

	origWD, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = os.Chdir(origWD) })

	base := t.TempDir()
	dirA := filepath.Join(base, "project-a")
	dirB := filepath.Join(base, "project-b")
	if err := os.MkdirAll(dirA, 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.MkdirAll(dirB, 0o755); err != nil {
		t.Fatal(err)
	}

	if err := os.Chdir(dirA); err != nil {
		t.Fatal(err)
	}
	pathA, err := lytecache.DefaultPath()
	if err != nil {
		t.Fatal(err)
	}
	pathAAgain, err := lytecache.DefaultPath()
	if err != nil {
		t.Fatal(err)
	}
	if pathA != pathAAgain {
		t.Fatalf("expected DefaultPath to be stable for the same cwd: %q vs %q", pathA, pathAAgain)
	}

	if err := os.Chdir(dirB); err != nil {
		t.Fatal(err)
	}
	pathB, err := lytecache.DefaultPath()
	if err != nil {
		t.Fatal(err)
	}

	if pathA == pathB {
		t.Fatalf("expected different default paths for different working directories, both got %q", pathA)
	}
}

File: lytecache-php/tests/Feature/ZeroConfigTest.php
<?php

declare(strict_types=1);

namespace Lytecache\Tests\Feature;

use Lytecache\LyteCache;
use Lytecache\Tests\TestCase;

final class ZeroConfigTest extends TestCase
{
    private ?string $originalHome = null;

    private ?string $originalXdg = null;

    private ?string $originalLytecachePath = null;

    protected function setUp(): void
    {
        parent::setUp();
        $this->originalHome = getenv('HOME') ?: null;
        $this->originalXdg = getenv('XDG_CACHE_HOME') ?: null;
        $this->originalLytecachePath = getenv('LYTECACHE_PATH') ?: null;
    }

    protected function tearDown(): void
    {
        $this->restoreEnv('HOME', $this->originalHome);
        $this->restoreEnv('XDG_CACHE_HOME', $this->originalXdg);
        $this->restoreEnv('LYTECACHE_PATH', $this->originalLytecachePath);
        parent::tearDown();
    }

    private function restoreEnv(string $name, ?string $value): void
    {
        if ($value === null) {
            putenv($name);
        } else {
            putenv("{$name}={$value}");
        }
    }

    public function test_zero_config_creates_file_and_parent_dirs(): void
    {
        $home = sys_get_temp_dir().'/lytecache-home-'.bin2hex(random_bytes(8));
        mkdir($home, 0755, true);
        putenv("HOME={$home}");
        putenv('XDG_CACHE_HOME');

        $path = LyteCache::defaultPath();
        self::assertFileDoesNotExist($path);

        $cache = new LyteCache(path: $path);
        self::assertFileExists($path);
        self::assertSame($path, $cache->path());

        $cache->close();
        $this->removeDirRecursive($home);
    }

    public function test_lytecache_path_env_override(): void
    {
        $dir = sys_get_temp_dir().'/lytecache-override-'.bin2hex(random_bytes(8));
        $override = $dir.'/custom/nested/path.db';
        putenv("LYTECACHE_PATH={$override}");

        self::assertSame($override, LyteCache::defaultPath());

        $cache = new LyteCache;
        self::assertSame($override, $cache->path());
        self::assertFileExists($override);

        $cache->close();
        $this->removeDirRecursive($dir);
    }

    public function test_different_working_directories_resolve_different_files(): void
    {
        $home = sys_get_temp_dir().'/lytecache-home-'.bin2hex(random_bytes(8));
        mkdir($home, 0755, true);
        putenv("HOME={$home}");
        putenv('XDG_CACHE_HOME');
        putenv('LYTECACHE_PATH');

        $base = sys_get_temp_dir().'/lytecache-cwd-'.bin2hex(random_bytes(8));
        $dirA = $base.'/project-a';
        $dirB = $base.'/project-b';
        mkdir($dirA, 0755, true);
        mkdir($dirB, 0755, true);

        $originalCwd = getcwd();

        chdir($dirA);
        $pathA = LyteCache::defaultPath();
        $pathAAgain = LyteCache::defaultPath();
        self::assertSame($pathA, $pathAAgain);

        chdir($dirB);
        $pathB = LyteCache::defaultPath();

        chdir((string) $originalCwd);

        self::assertNotSame($pathA, $pathB);

        $this->removeDirRecursive($home);
        $this->removeDirRecursive($base);
    }

    private function removeDirRecursive(string $dir): void
    {
        if (! is_dir($dir)) {
            return;
        }

        $items = scandir($dir) ?: [];
        foreach ($items as $item) {
            if ($item === '.' || $item === '..') {
                continue;
            }
            $path = $dir.'/'.$item;
            is_dir($path) ? $this->removeDirRecursive($path) : @unlink($path);
        }

        @rmdir($dir);
    }
}
